tsdate/core.py

```diff
@@ -226,40 +228,43 @@ def get_modified_ts(self, result, eps):
             )
         return tables.tree_sequence()
 
-    def set_time_metadata(self, table, mean, var, default_schema, overwrite=False):
-        if var is not None:
-            table_name = type(table).__name__
-            assert len(mean) == len(var) == table.num_rows
-            if table.metadata_schema.schema is None or overwrite:
-                if len(table.metadata) == 0 or overwrite:
-                    table.metadata_schema = default_schema
-                    md_iter = ({} for _ in range(table.num_rows))
-                    # For speed, assume we don't need to validate
-                    encoder = table.metadata_schema.encode_row
-                    logger.info(f"Set metadata schema on {table_name}")
-                else:
-                    logger.warning(
-                        f"Could not set metadata on {table_name}: "
-                        "data already exists with no schema"
-                    )
-                    return
+    def set_time_metadata(self, table, mean, var, default_schema):
+        # Try to set metadata: if we fail, clear metadata, reset schema, and try again
+        def _time_md_array(table, mean, var):
+            # Return an array of metadata dicts, or raise an error if
+            # schema is None or metadata is not valid
+            s = table.metadata_schema
+            if s.schema is None:
+                raise tskit.MetadataEncodingError("No schema set")
+            encoder = s.encode_row if s == default_schema else s.validate_and_encode_row
```

Comment:
I'm not quite getting why we have to have the `if` here? Why does default schema not have `validate_and_encode_row`?

Reply:
Oh, good point. I put that in because I didn't want the hassle of validating json metadata. but actually this would probably not cause efficiency issues, so I guess we can simply call `validate_and_encode_row` for json data too?

Reply:
Yeah, I'd rather have a single code path here unless and until we have a proven bottleneck.